server/controllers/chartsController/index.js

```diff
@@ -122,35 +122,92 @@ const getAllSubjects = async ({ dataDb, chart, userAccess, filters }) => {
       )
       .toArray()
   } else {
-    const { mongoAggregateQueryForFilters, activeFilters } =
-      mongoQueryFromFilters(filters)
+    const {
+      mongoAggregateQueryForIncludedCriteria,
+      mongoAggregateQueryForFilters,
+      mongoQueryForSocioDemographics,
+      activeFilters,
+    } = mongoQueriesFromFilters(filters, userAccess)
+    const mapForResultsFromFilters = new Map()
     const filteredSubjects = []
-    const criteriaCursor = await dataDb.collection(collections.toc).find(
-      {
-        assessment: INCLUSION_EXCLUSION_CRITERIA_FORM,
-        study: { $in: userAccess, $nin: STUDIES_TO_OMIT },
-      },
-      {
-        projection: ALL_SUBJECTS_MONGO_PROJECTION,
-      }
-    )
+    const criteriaCursor = await dataDb
+      .collection(collections.toc)
+      .aggregate(mongoAggregateQueryForFilters)
 
     for await (let doc of criteriaCursor) {
-      const data = await dataDb
-        .collection(doc.collection)
-        .aggregate(mongoAggregateQueryForFilters)
-      const subjectCriteriaData = await data.next()
-      const isSubjectInFilteredQuery = activeFilters
-        .map(
-          (requestedFilter) => subjectCriteriaData[requestedFilter].length > 0
-        )
-        .every(Boolean)
-
-      if (isSubjectInFilteredQuery) {
-        filteredSubjects.push(doc.subject)
+      if (doc.socioDemographics && doc.inclusionCriteria) {
+        if (doc.socioDemographics.length <= doc.inclusionCriteria.length) {
+          addSubjectsToResultsFromFiltersMap({
+            subjectsInFiltersA: doc.socioDemographics,
+            subjectsInFiltersB: doc.inclusionCriteria,
+            mapForResultsFromFilters,
+          })
+        } else if (
+          doc.inclusionCriteria.length < doc.socioDemographics.length
+        ) {
+          addSubjectsToResultsFromFiltersMap({
+            subjectsInFiltersA: doc.inclusionCriteria,
+            subjectsInFiltersB: doc.socioDemographics,
+            mapForResultsFromFilters,
+          })
+        }
+      } else {
+        if (doc.socioDemographics?.length && !doc.inclusionCriteria) {
+          doc.socioDemographics.forEach((data) => {
+            if (mapForResultsFromFilters.has(data.subject)) {
+              return false
+            }
+            mapForResultsFromFilters.set(data.subject, data)
+          })
+        } else if (doc.inclusionCriteria?.length && !doc.socioDemographics) {
+          doc.inclusionCriteria.forEach((data) => {
+            if (mapForResultsFromFilters.has(data.subject)) {
+              return false
+            }
+            mapForResultsFromFilters.set(data.subject, data)
+          })
+        }
```

Comment:
I think we can clean up these nested conditionals to make the code more readable.

Reply:
https://github.com/AMP-SCZ/dpdash/pull/280/commits/c30a7d514c931ad36f83916c440ff7812449f315